### PR TITLE
Timing API implementation

### PR DESCRIPTION
**Pull request by jsconan**
Related to: https://oat-sa.atlassian.net/browse/TAO-2487

Requires: 
- https://github.com/oat-sa/extension-tao-test/pull/89
- https://github.com/oat-sa/extension-tao-test/pull/90
- https://github.com/oat-sa/extension-tao-test/pull/91

### CODE REVIEW

File: models/classes/runner/time/QtiTimeLine.php [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+     */
+    public function serialize()
+    {
+        $data = [];
+
+        foreach ($this->points as $point) {
+            $data [] = serialize($point);
+        }
+
+        return serialize($data);
```
Comment (krampstudio) [inline models/classes/runner/time/QtiTimeLine.php]: Why do you iterate over the points ? Isn't `serialize($this->points)` sufficient ?

File: models/classes/runner/time/QtiTimeLine.php [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    {
+        // the tag criteria can be omitted
+        $tags = null;
+        if (isset($tag)) {
+            $tags = is_array($tag) ? $tag : [$tag];
+        }
+        
+        // create a another instance of the same class
+        $class = get_class($this);
+        $subset = new $class();
```
Comment (krampstudio) [inline models/classes/runner/time/QtiTimeLine.php]: not sure, but If I remember there's something like `new static()` introduced in php ~5.2 to create a new instance of the current class safely. (please verfiy)

File: models/classes/runner/time/QtiTimer.php [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+            throw new InvalidDataException('start() needs a valid timestamp!');
+        }
+        
+        // extract the TimePoint identification from the provided item, and find existing range
+        $tags = $this->getItemTags($routeItem);
+        $range = $this->timeLine->find($tags, TimePoint::TARGET_SERVER);
+
+        // validate the data consistence
+        if (count($range) % 2) {
+            throw new InconsistentRangeException('The time range does not seem to be consistent, the range is not complete!');
```
Comment (krampstudio) [inline models/classes/runner/time/QtiTimer.php]: I'm not sure about that one, what if the browser crash. You'll have a start without end, or is it going to be a new occurence ?

File: models/classes/runner/time/QtiTimer.php [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+
+        $occurrence = $routeItem->getOccurence();
+        $partId = $testPart->getIdentifier();
+        $sectionId = key(current($sections));
+        $itemId = $itemRef->getIdentifier();
+        
+        $tags []= $itemId;
+        $tags []= $itemId . '#' . $occurrence;
+        $tags []= $partId;
+        $tags []= $sectionId;
```
Comment (krampstudio) [inline models/classes/runner/time/QtiTimer.php]: We miss the testId tag ? or how do we compute the timeout for the wall test ?

Comment (jsconan) [inline models/classes/runner/time/QtiTimer.php]: A time line is related to a test, so if we compute the duration without any criteria, the result will embrace the whole test.

**Pull request merged**